extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/PollingLookup.java

```diff
@@ -0,0 +1,242 @@
+      throw new ISE("Cache reference is null WTF");
+    }
+    PollingCache cache = cacheRefKeeper.getAndIncrementRef();
+    try {
+      if (cache == null) {
+        // it must've been closed after swapping while I was getting it.  Try again.
+        return this.apply(key);
+      }
+      return Strings.emptyToNull((String) cache.get(key));
+    }
+    finally {
+      if (cacheRefKeeper != null && cache != null) {
+        cacheRefKeeper.doneWithIt();
+      }
+    }
+  }
+
+  @Override
+  public List<String> unapply(final String value)
+  {
+    CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.get();
+    if (cacheRefKeeper == null) {
+      throw new ISE("pollingLookup id [%s] is closed", id);
+    }
+    PollingCache cache = cacheRefKeeper.getAndIncrementRef();
+    try {
+      if (cache == null) {
+        // it must've been closed after swapping while I was getting it.  Try again.
+        return this.unapply(value);
+      }
+      return cache.getKeys(value);
+    }
+    finally {
+      if (cacheRefKeeper != null && cache != null) {
+        cacheRefKeeper.doneWithIt();
+      }
+    }
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    return LookupExtractionModule.getRandomCacheKey();
+  }
+
+  private Runnable pollAndSwap()
+  {
+    return new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        LOGGER.debug("Polling and swapping of PollingLookup [%s]", id);
+        CacheRefKeeper newCacheKeeper = new CacheRefKeeper(cacheFactory.makeOf(dataFetcher.fetchAll()));
+        CacheRefKeeper oldCacheKeeper = refOfCacheKeeper.getAndSet(newCacheKeeper);
+        if (oldCacheKeeper != null) {
+          oldCacheKeeper.doneWithIt();
+        }
+      }
+    };
+  }
+
+  @Override
+  public int hashCode()
+  {
+    int result = (int) (pollPeriodMs ^ (pollPeriodMs >>> 32));
+    result = 31 * result + dataFetcher.hashCode();
+    result = 31 * result + cacheFactory.hashCode();
+    return result;
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof PollingLookup)) {
+      return false;
+    }
+
+    PollingLookup that = (PollingLookup) o;
+
+    if (pollPeriodMs != that.pollPeriodMs) {
+      return false;
+    }
+    if (!dataFetcher.equals(that.dataFetcher)) {
+      return false;
+    }
+    return cacheFactory.equals(that.cacheFactory);
+
+  }
+
+  public boolean isOpen()
+  {
+    return isOpen.get();
+  }
+
+
+  protected static class CacheRefKeeper
+  {
+    private final PollingCache pollingCache;
+    private volatile Long refCounts = 0L;
+
+    CacheRefKeeper(PollingCache pollingCache) {this.pollingCache = pollingCache;}
+
+    PollingCache getAndIncrementRef()
+    {
+      synchronized (refCounts) {
+        if (refCounts < 0) {
+          return null;
+        }
+        refCounts += 1;
+        return pollingCache;
+      }
+    }
+
+    void doneWithIt()
+    {
+      synchronized (refCounts) {
```

Comment:
do `-=` or `+=` modify a `Long` boxing in place or does it replace the boxed object?

That's a pretty gray area, and I highly suggest synchronizing on a guaranteed fixed object unless there are sufficient unit tests or specification reference to show that a `Long` boxing modifies in place with `-=` and `+=`.